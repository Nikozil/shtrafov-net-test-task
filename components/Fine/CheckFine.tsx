import { NextPage } from 'next';
import React, { ChangeEvent, FormEvent, useState } from 'react';
import { getFine } from '../../api/api';
import Loader from '../../assets/Loader/Loader';
import getHelpUin from '../../lib/getHelpUin';
import styles from '../../styles/CheckFine.module.scss';
import FineFound from './FineFound';
import FineNotFound from './FineNotFound';
import Form from './Form';

const CheckFine: NextPage = () => {
  const [isStartSearch, setIsStartSearch] = useState<boolean>(false);
  const [status, setStatus] = useState<string | null>(null);
  const [isSearching, setIsSearching] = useState<boolean>(false);
  const [fine, setFine] = useState<Fine | null | false>(null);
  const [uin, setUin] = useState<UIN>('');
  const [helperUin, setHelperUin] = useState<HelperUIN>(null);

  const handleSubmit = (e: FormEvent) => {
    e.preventDefault();
    setIsStartSearch(true);
    setFine(null);

    const fetchFine = async () => {
      setIsSearching(true);
      let responce = await getFine(uin);
      !responce && setStatus('Неверные данные');
      setFine(responce);
      setIsSearching(false);
    };
    uin.length === 25 || uin.length === 20
      ? fetchFine()
      : setStatus('Количество символов должно быть 20 или 25');
  };
  const handleChange = (e: ChangeEvent<HTMLInputElement>) => {
    const value = e.target.value;
    setIsStartSearch(false);
    setUin(value);
    const helpUin = getHelpUin(value);

    setHelperUin(helpUin);
  };
  const handleHelpClick = () => {
    if (helperUin) {
      setUin(helperUin);
      setHelperUin(null);
    }
  };

  return (
    <div className={styles.content}>
      <div>Получение информации о штрафе по УИН</div>
      <Form
        uin={uin}
        helperUin={helperUin}
        handleSubmit={handleSubmit}
        handleChange={handleChange}
        handleHelpClick={handleHelpClick}
      />
      {isStartSearch ? (
        isSearching ? (
          <Loader />
        ) : fine ? (
          <FineFound fine={fine} />
        ) : fine === false ? (
          <FineNotFound uin={uin} />
        ) : (
          <div className={styles.status}> {status}</div>
        )
      ) : null}
    </div>
  );
};

export default CheckFine;

export type UIN = string;
export type HelperUIN = string | null;
export interface Fine {
  amount: string | null;
  amount_to_pay: string | null;
  bill_at: string | null;
  discount_size: string | null;
  division_name: string | null;
  doc_number: string | null;
  doc_type: string | null;
  kbk: string | null;
  koap_code: string | null;
  name: string | null;
  number: string | null;
  pay_status: string | null;
  payee_bank_account: string | null;
  payee_bank_bik: string | null;
  payee_bank_name: string | null;
  payee_inn: string | null;
  payee_kpp: string | null;
  payee_oktmo: string | null;
  payee_username: string | null;
  payment_status: string | null;
  quittance: string | null;
  violation_at: string | null;
  violator_name: string | null;
}
