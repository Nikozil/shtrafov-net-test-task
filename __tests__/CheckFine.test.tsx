import { render, waitFor } from '@testing-library/react';
import React from 'react';
import CheckFine from '../components/Fine/CheckFine';

describe('CheckFine', () => {
  it('renders unchanged', async () => {
    const finecomp = render(<CheckFine />);
    await waitFor(() => {
      expect(finecomp).toMatchSnapshot();
    });
  });
});
