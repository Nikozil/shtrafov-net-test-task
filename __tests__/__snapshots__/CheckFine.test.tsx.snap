// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckFine renders unchanged 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="content"
      >
        <div>
          Получение информации о штрафе по УИН
        </div>
        <form
          class="search-form"
        >
          <label
            class="search-form__label"
            for="UIN"
          >
            Введите УИН
          </label>
          <input
            autocomplete="off"
            class="search-form__input"
            id="UIN"
            placeholder="Введите УИН"
            value=""
          />
          <button
            class="search-form__button"
          >
            Найти
          </button>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="content"
    >
      <div>
        Получение информации о штрафе по УИН
      </div>
      <form
        class="search-form"
      >
        <label
          class="search-form__label"
          for="UIN"
        >
          Введите УИН
        </label>
        <input
          autocomplete="off"
          class="search-form__input"
          id="UIN"
          placeholder="Введите УИН"
          value=""
        />
        <button
          class="search-form__button"
        >
          Найти
        </button>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
